File: 25. Reverse Nodes in k-Group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    var dummy = new ListNode(0);
    dummy.next = head;

    head = dummy;
    while( true ) {
        head = reverseK( head, k );
        if( head == null ) {
            break;
        }
    }

    return dummy.next;

    // head -> n1 -> n2 -> n3 -> n4 -> ..... -> nk -> nk+1
    // head -> nk -> nk-1 -> nk-2 -> ..... -> n2 -> n1 -> nk+1
    // return n1
    function reverseK( head, k ) {
        var nk = head;
        for ( var i = 0; i < k; i++ ) {
            if ( nk == null ) {
                return null;
            }
            nk = nk.next;
        }

        if ( nk == null ) {
            return null;
        }

        // reverse
        var n1 = head.next;
        var nkplus = nk.next;

        var prev = null;
        var curt = n1;

        while ( curt != nkplus ) {
            var tmp = curt.next;
            curt.next = prev;
            prev = curt;
            curt = tmp;
        }

        // connect
        head.next = nk;
        n1.next = nkplus;
        return n1;
    }
};

File: 25. Reverse Nodes in k-Group.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head is None:
            return
        dummyNode = ListNode(-1)
        dummyNode.next = head
        head = dummyNode
        while True:
            head = self.helper(head, k)
            if head is None:
                break
        return dummyNode.next

    # head -> n1 -> n2 -> n3 -> n4 -> n5 -> n6 -> ..... -> nk -> nk+1
    def helper(self, head, k):
        nk = head
        for i in xrange(k):
            if nk is None:
                return None
            nk = nk.next
        if nk is None:
            return None
        n1 = head.next

        while head.next != nk:
            tmp = n1.next.next
            n1.next.next = head.next
            head.next = n1.next
            n1.next = tmp
        return n1
